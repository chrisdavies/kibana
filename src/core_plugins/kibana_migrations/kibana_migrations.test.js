const _ = require('lodash');
const sinon = require('sinon');
const migrationsPlugin = require('./index');
const { mockCluster } = require('../../../packages/kbn-migrations/src/test');

describe('kibana_migrations plugin', () => {
  describe('init', () => {

    test('goes from yellow to green', async () => {
      const { plugin, kbnServer } = await testMigrationPlugin();

      expect(kbnServer.server.plugins.elasticsearch.waitUntilReady.calledOnce)
        .toBeTruthy();
      expect(plugin.status.yellow.args).toEqual([
        ['Waiting for elasticsearch...'],
        ['Migrating the Kibana index...'],
      ]);
      expect(plugin.status.value).toEqual('green');
    });

    test('migrates the kibana index', async () => {
      const pluginSpecs = [{
        id: 'p1',
        mappings: {
          field1: { type: 'text' },
        },
        migrations: [{
          id: 'mg1',
          seed: () => ({ id: 'a', type: 'field1', attributes: 'Dangit' }),
        }],
      }, {
        id: 'p2',
        mappings: {
          field2: { type: 'text' },
        },
        migrations: [{
          id: 'mg2',
          seed: () => ({ id: 'b', type: 'field2', attributes: 'Bobby' }),
        }],
      }];
      const adminCluster = mockCluster({}, {});
      await testMigrationPlugin({ adminCluster, pluginSpecs });

      expect(adminCluster.state()).toMatchSnapshot();
    });

    test('waits for the index to migrate', async () => {
      let timedOut = false;
      const data = {
        '.kibana': {
          'migration:migration-state': {
            _source: {
              migration: {
                status: 'migrating',
                plugins: [],
              },
            },
          },
        },
      };
      const adminCluster = mockCluster(data, {});
      const kbnServer = mockKbnServer({ adminCluster });
      const plugin = migrationsPlugin(mockKibana(kbnServer));
      const promise = plugin.init(kbnServer.server);

      setTimeout(() => {
        expect(plugin.status.value).toEqual('yellow');
        _.set(data, ['.kibana', 'migration:migration-state', '_source', 'migration', 'status'], 'migrated');
        timedOut = true;
      }, 20);

      await promise;
      expect(timedOut).toBeTruthy();
      expect(plugin.status.value).toEqual('green');
    });

    test('respects the force option', async () => {
      const data = {
        '.kibana': {
          'migration:migration-state': {
            _source: {
              migration: {
                status: 'migrating',
                plugins: [],
              },
            },
          },
        },
      };
      const adminCluster = mockCluster(data, {});
      const { plugin } = await testMigrationPlugin({ adminCluster, force: true });

      expect(plugin.status.value).toEqual('green');
      expect(adminCluster.state()).toMatchSnapshot();
    });
  });

  describe('migrationOptions', () => {
    test('exposes whatever callCluster it is passed', async () => {
      const callCluster = () => {};
      const { kbnServer } = await testMigrationPlugin();
      const opts = kbnServer.server.exposed.migrationOptions({ callCluster });
      expect(opts.callCluster).toEqual(callCluster);
    });

    test('exposes callWithInternalUser if no callCluster is specified', async () => {
      const adminCluster = () => {};
      const { kbnServer } = await testMigrationPlugin({ adminCluster });
      const opts = kbnServer.server.exposed.migrationOptions();
      expect(opts.callCluster).toEqual(adminCluster);
    });

    test('exposes configured index', async () => {
      const index = Math.random().toString();
      const { kbnServer } = await testMigrationPlugin({ index });
      const opts = kbnServer.server.exposed.migrationOptions();
      expect(opts.index).toEqual(index);
    });

    test('exposes elastic version', async () => {
      const version = Math.random().toString();
      const { kbnServer } = await testMigrationPlugin({ version });
      const opts = kbnServer.server.exposed.migrationOptions();
      expect(opts.elasticVersion).toEqual(version);
    });

    test('exposes the server log function', async () => {
      const { kbnServer } = await testMigrationPlugin();
      const opts = kbnServer.server.exposed.migrationOptions();
      opts.log(['hello'], 'world');
      expect(_.last(kbnServer.server.log.args)).toEqual([['hello'], 'world']);
    });

    test('exposes pluginSpecs', async () => {
      const pluginSpecs = [{
        id: 'foo',
        mappings: {
          bar: { type: 'text' },
        },
        migrations: [{
          id: 'mg1',
          seed: () => ({ type: 'bar', attributes: 'Dangit' }),
        }],
      }, {
        id: 'sans-stuff',
        mappings: undefined,
        migrations: undefined,
      }];
      const adminCluster = mockCluster({}, {});
      const { kbnServer } = await testMigrationPlugin({ pluginSpecs, adminCluster });
      const opts = kbnServer.server.exposed.migrationOptions();
      expect(opts.plugins).toEqual(pluginSpecs);
    });
  });
});

async function testMigrationPlugin(config) {
  const kbnServer = mockKbnServer(config);
  const plugin = migrationsPlugin(mockKibana(kbnServer));
  await plugin.init(kbnServer.server);

  return { plugin, kbnServer };
}

function mockKibana(kbnServer) {
  function Plugin({ init }) {
    this.kbnServer = kbnServer || mockKbnServer();
    this.status = {
      yellow: sinon.spy(() => this.status.value = 'yellow'),
      green: sinon.spy(() => this.status.value = 'green'),
    };
    this.init = init;
  }

  return { Plugin };
}

function mockKbnServer(config = {}) {
  return {
    version: config.version || '4.5.6',
    server: mockServer(config),
    settings: {
      migration: {
        force: config.force,
      },
    },
    pluginSpecs: (config.pluginSpecs || []).map((pluginSpec) => ({
      getId: () => pluginSpec.id,
      getExportSpecs: () => _.cloneDeep(pluginSpec),
      getMigrations: () => _.cloneDeep(pluginSpec.migrations),
    })),
  };
}

function mockServer({ adminCluster, index } = {}) {
  const clusterMap = {
    admin: {
      callWithInternalUser: adminCluster || sinon.spy()
    },
  };
  const configMap = {
    'kibana.index': index || '.kibana',
    'kibanamigrations.pollInterval': 10,
  };
  const server = {
    exposed: {},
    expose: sinon.spy((k, v) => server.exposed[k] = v),
    config() {
      return {
        get(key) {
          return configMap[key];
        }
      };
    },
    log: sinon.spy(),
    plugins: {
      elasticsearch: {
        getCluster(name) {
          return clusterMap[name];
        },
        waitUntilReady: sinon.spy(() => Promise.resolve()),
      },
    },
  };

  return server;
}
