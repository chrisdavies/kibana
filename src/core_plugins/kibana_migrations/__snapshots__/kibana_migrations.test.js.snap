// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kibana_migrations plugin init migrates the kibana index 1`] = `
Object {
  "data": Object {
    ".kibana-4.5.6-438e8cb14dfd248f7f3898d52c746daba6f47fbe": Object {
      "field1:a": Object {
        "_source": Object {
          "field1": "Dangit",
          "type": "field1",
          "updated_at": undefined,
        },
      },
      "field2:b": Object {
        "_source": Object {
          "field2": "Bobby",
          "type": "field2",
          "updated_at": undefined,
        },
      },
      "migration:migration-state": Object {
        "_source": Object {
          "migration": Object {
            "plugins": Array [
              Object {
                "id": "p1",
                "mappings": "{\\"field1\\":{\\"type\\":\\"text\\"}}",
                "mappingsChecksum": "ebe9241e1163529656d3c2174d0a5d35928d7fff",
                "migrationIds": Array [
                  "mg1",
                ],
                "migrationsChecksum": "3de7d9b63b0a153410b3f636669de57e9149ad54",
              },
              Object {
                "id": "p2",
                "mappings": "{\\"field2\\":{\\"type\\":\\"text\\"}}",
                "mappingsChecksum": "08d8c7375209c783055f100449d53911c72c522b",
                "migrationIds": Array [
                  "mg2",
                ],
                "migrationsChecksum": "ad525bb6c4f31b0442e08963e9052e87ebc46cc7",
              },
            ],
            "previousIndex": ".kibana-4.5.6-original",
            "status": "migrated",
          },
          "type": "migration",
        },
        "_version": 1,
      },
    },
    ".kibana-4.5.6-original": Object {
      "migration:migration-state": Object {
        "_source": Object {
          "migration": Object {
            "plugins": Array [],
            "status": "migrating",
          },
          "type": "migration",
        },
        "_version": 1,
      },
    },
  },
  "meta": Object {
    "aliases": Object {
      ".kibana": Object {
        ".kibana-4.5.6-438e8cb14dfd248f7f3898d52c746daba6f47fbe": Object {
          "aliases": Object {
            ".kibana": Object {},
          },
        },
      },
    },
    "mappings": Object {
      ".kibana-4.5.6-438e8cb14dfd248f7f3898d52c746daba6f47fbe": Object {
        "doc": Object {
          "dynamic": "strict",
          "properties": Object {
            "config": Object {
              "dynamic": true,
              "properties": Object {
                "buildNum": Object {
                  "type": "keyword",
                },
              },
            },
            "field1": Object {
              "type": "text",
            },
            "field2": Object {
              "type": "text",
            },
            "migration": Object {
              "properties": Object {
                "plugins": Object {
                  "properties": Object {
                    "id": Object {
                      "type": "keyword",
                    },
                    "mappings": Object {
                      "type": "text",
                    },
                    "mappingsChecksum": Object {
                      "type": "keyword",
                    },
                    "migrationIds": Object {
                      "type": "keyword",
                    },
                    "migrationsChecksum": Object {
                      "type": "keyword",
                    },
                  },
                  "type": "nested",
                },
                "previousIndex": Object {
                  "type": "keyword",
                },
                "status": Object {
                  "type": "keyword",
                },
              },
            },
            "type": Object {
              "type": "keyword",
            },
            "updated_at": Object {
              "type": "date",
            },
          },
        },
      },
      ".kibana-4.5.6-original": Object {
        "doc": Object {
          "dynamic": "strict",
          "properties": Object {
            "config": Object {
              "dynamic": true,
              "properties": Object {
                "buildNum": Object {
                  "type": "keyword",
                },
              },
            },
            "field1": Object {
              "type": "text",
            },
            "field2": Object {
              "type": "text",
            },
            "migration": Object {
              "properties": Object {
                "plugins": Object {
                  "properties": Object {
                    "id": Object {
                      "type": "keyword",
                    },
                    "mappings": Object {
                      "type": "text",
                    },
                    "mappingsChecksum": Object {
                      "type": "keyword",
                    },
                    "migrationIds": Object {
                      "type": "keyword",
                    },
                    "migrationsChecksum": Object {
                      "type": "keyword",
                    },
                  },
                  "type": "nested",
                },
                "previousIndex": Object {
                  "type": "keyword",
                },
                "status": Object {
                  "type": "keyword",
                },
              },
            },
            "type": Object {
              "type": "keyword",
            },
            "updated_at": Object {
              "type": "date",
            },
          },
        },
      },
    },
    "settings": Object {
      ".kibana-4.5.6-438e8cb14dfd248f7f3898d52c746daba6f47fbe": Object {
        "index": Object {
          "number_of_replicas": "1",
          "number_of_shards": "5",
        },
      },
      ".kibana-4.5.6-original": Object {
        "index": Object {
          "blocks.read_only": false,
          "number_of_replicas": "1",
          "number_of_shards": "5",
        },
      },
    },
  },
}
`;

exports[`kibana_migrations plugin init respects the force option 1`] = `
Object {
  "data": Object {
    ".kibana-4.5.6-989db2448f309bfdd99b513f37c84b8f5794d2b5": Object {
      "migration:migration-state": Object {
        "_source": Object {
          "migration": Object {
            "plugins": Array [],
            "previousIndex": ".kibana-4.5.6-original",
            "status": "migrated",
          },
          "type": "migration",
        },
        "_version": 1,
      },
    },
    ".kibana-4.5.6-original": Object {
      "migration:migration-state": Object {
        "_source": Object {
          "migration": Object {
            "plugins": Array [],
            "status": "migrating",
          },
          "type": "migration",
        },
        "_version": 1,
      },
    },
  },
  "meta": Object {
    "aliases": Object {
      ".kibana": Object {
        ".kibana-4.5.6-989db2448f309bfdd99b513f37c84b8f5794d2b5": Object {
          "aliases": Object {
            ".kibana": Object {},
          },
        },
      },
    },
    "mappings": Object {
      ".kibana-4.5.6-989db2448f309bfdd99b513f37c84b8f5794d2b5": Object {
        "doc": Object {
          "dynamic": "strict",
          "properties": Object {
            "config": Object {
              "dynamic": true,
              "properties": Object {
                "buildNum": Object {
                  "type": "keyword",
                },
              },
            },
            "migration": Object {
              "properties": Object {
                "plugins": Object {
                  "properties": Object {
                    "id": Object {
                      "type": "keyword",
                    },
                    "mappings": Object {
                      "type": "text",
                    },
                    "mappingsChecksum": Object {
                      "type": "keyword",
                    },
                    "migrationIds": Object {
                      "type": "keyword",
                    },
                    "migrationsChecksum": Object {
                      "type": "keyword",
                    },
                  },
                  "type": "nested",
                },
                "previousIndex": Object {
                  "type": "keyword",
                },
                "status": Object {
                  "type": "keyword",
                },
              },
            },
            "type": Object {
              "type": "keyword",
            },
            "updated_at": Object {
              "type": "date",
            },
          },
        },
      },
      ".kibana-4.5.6-original": Object {
        "doc": Object {
          "properties": Object {
            "config": Object {
              "dynamic": true,
              "properties": Object {
                "buildNum": Object {
                  "type": "keyword",
                },
              },
            },
            "migration": Object {
              "properties": Object {
                "plugins": Object {
                  "properties": Object {
                    "id": Object {
                      "type": "keyword",
                    },
                    "mappings": Object {
                      "type": "text",
                    },
                    "mappingsChecksum": Object {
                      "type": "keyword",
                    },
                    "migrationIds": Object {
                      "type": "keyword",
                    },
                    "migrationsChecksum": Object {
                      "type": "keyword",
                    },
                  },
                  "type": "nested",
                },
                "previousIndex": Object {
                  "type": "keyword",
                },
                "status": Object {
                  "type": "keyword",
                },
              },
            },
            "type": Object {
              "type": "keyword",
            },
            "updated_at": Object {
              "type": "date",
            },
          },
        },
      },
    },
    "settings": Object {
      ".kibana-4.5.6-989db2448f309bfdd99b513f37c84b8f5794d2b5": Object {
        "index": Object {
          "number_of_replicas": undefined,
          "number_of_shards": undefined,
        },
      },
      ".kibana-4.5.6-original": Object {
        "index": Object {
          "blocks.read_only": false,
          "number_of_replicas": undefined,
          "number_of_shards": undefined,
        },
      },
    },
  },
}
`;
